cmd/jaeger/integration/receivers/storagereceiver/receiver_test.go

```diff
@@ -0,0 +1,264 @@
+
+	r, err := newTracesReceiver(
+		cfg,
+		receivertest.NewNopCreateSettings(),
+		consumertest.NewNop(),
+	)
+	require.NoError(t, err)
+
+	err = r.Start(context.Background(), componenttest.NewNopHost())
+	require.ErrorContains(t, err, "cannot find storage factory")
+}
+
+func TestReceiverStart(t *testing.T) {
+	ctx := context.Background()
+	host := newStorageHost(t, "external-storage")
+
+	cfg := createDefaultConfig().(*Config)
+	cfg.TraceStorage = "external-storage"
+
+	r, err := newTracesReceiver(
+		cfg,
+		receivertest.NewNopCreateSettings(),
+		consumertest.NewNop(),
+	)
+	require.NoError(t, err)
+
+	require.NoError(t, r.Start(ctx, host))
+	require.NoError(t, r.Shutdown(ctx))
+}
+
+func TestReceiverStartConsume(t *testing.T) {
+	sink := &consumertest.TracesSink{}
+
+	cfg := createDefaultConfig().(*Config)
+	cfg.TraceStorage = "external-storage"
+
+	r, _ := newTracesReceiver(cfg, receivertest.NewNopCreateSettings(), sink)
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	r.cancelConsumeLoop = cancelFunc
+
+	tests := []struct {
+		name           string
+		services       []string
+		traces         []*model.Trace
+		tracesErr      error
+		expectedTraces []*model.Trace
+	}{
+		{
+			name: "empty service",
+		},
+		{
+			name:      "find traces error",
+			services:  []string{"example-service"},
+			tracesErr: errors.New("failed to find traces"),
+		},
+		{
+			name:     "consume first trace",
+			services: []string{services[0]},
+			traces: []*model.Trace{
+				{Spans: []*model.Span{spans[0]}},
+			},
+			expectedTraces: []*model.Trace{
+				{Spans: []*model.Span{spans[0]}},
+			},
+		},
+		{
+			name:     "consume second trace",
+			services: services,
+			traces: []*model.Trace{
+				{Spans: []*model.Span{spans[0]}},
+				{Spans: []*model.Span{spans[2], spans[3]}},
+			},
+			expectedTraces: []*model.Trace{
+				{Spans: []*model.Span{spans[0]}},
+				{Spans: []*model.Span{spans[2]}},
+				{Spans: []*model.Span{spans[3]}},
+			},
+		},
+		{
+			name:     "re-consume first trace with new spans",
+			services: services,
+			traces: []*model.Trace{
+				{Spans: []*model.Span{spans[0], spans[1]}},
+				{Spans: []*model.Span{spans[2], spans[3]}},
+			},
+			expectedTraces: []*model.Trace{
+				{Spans: []*model.Span{spans[0]}},
+				{Spans: []*model.Span{spans[2]}},
+				{Spans: []*model.Span{spans[3]}},
+				// span at index 1 is consumed last
+				{Spans: []*model.Span{spans[1]}},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			reader := new(mocks.Reader)
+			reader.On("GetServices", mock.AnythingOfType("*context.cancelCtx")).Return(test.services, nil)
+			for _, service := range test.services {
+				reader.On(
+					"FindTraces",
+					mock.AnythingOfType("*context.cancelCtx"),
+					&spanstore.TraceQueryParameters{ServiceName: service},
+				).Return(test.traces, test.tracesErr)
+			}
+			r.spanReader = reader
+
+			require.NoError(t, r.Shutdown(ctx))
+			err := r.consumeLoop(ctx)
```

Comment:
are you sure you're actually testing different branches of accumulation of spans? Since you call consumeLoop after canceling the context, it will run exactly once. In other words, I don't really see how different test cases would actually result in different behavior, since some of the behavior is only triggered if the loop runs more than once.

Reply:
Yes, because I use the same receiver for every test case, the whole test sequence is a continuation from the previous case. Also, the behavior of calling consumeLoop after cancelling the context is expected. This approach allows me to control the simulation of consumeLoop, assuming that each test case represents an iteration of the for loop in the consumeLoop function and read the updated traces.